### ISSUE TITLE
sels_xgboost

### ISSUE THREAD

**Issue by refree** [refree/xgboost#1]
#xgboost 연습 _ target변수 형식은 오직 intiger만 모형에서 돌아감
#팩터, 뉴메릭 안돌아감.
#reference blog:http://jamesmarquezportfolio.com/get_up_and_running_with_xgboost_in_r.html

setwd("C:/Users/Park Jung/Desktop/SELS/data/query")
library(xgboost)
library(Matrix)
library(DiagrammeR)
library(dplyr)
 
  df <- read.csv("7grade_join_1.csv" )
  df <-as.data.frame(df)
  for(i in 1:length(df)) {   df[,i] <- as.numeric(df[,i]) }
 

  df <- df %>%
    mutate(do_violence_in_school = ifelse(do_violence_in_school >= 0.03965, 1, 0))
  df$do_violence_in_school <- as.integer(df$do_violence_in_school)

#-------------------------------------
library(data.table)
str(df)
df$gender = as.factor(df$gender)
df$residence_code = as.factor(df$residence_code)
#df$do_violence_in_school <- as.factor(df$do_violence_in_school)
library(caret)
trainIndex   <- createDataPartition(df$do_violence_in_school, p=0.75, list=FALSE, times=1)
train        <- df[ trainIndex, ]
test         <- df[-trainIndex, ]
str(df)
Train.X  <- sparse.model.matrix(do_violence_in_school ~ .-1, data=train)
Test.X   <- sparse.model.matrix(do_violence_in_school ~ .-1, data=test)

Train.Y <- train$do_violence_in_school
Test.Y <- test$do_violence_in_school

param <- list(objective   = "binary:logistic",
              eval_metric = "error",
              max_depth   = 7,
              eta         = 0.1,
              gammma      = 1,
              colsample_bytree = 0.5,
              min_child_weight = 1)

set.seed(1234)

#Pass in our hyperparameteres and train the model 
system.time(xgb <- xgboost(params  = param,
                           data    = Train.X,
                           label   = Train.Y, 
                           nrounds = 500,
                           print_every_n = 100,
                           verbose = 1))

xgb
#https://stackoverflow.com/questions/51607664/r-script-xgboost-for-binary-classification-how-to-get-predicted-label
#_혼동행렬 도출(7시간 만에 찾음)___________________________________________________________________________________
xgb_pred <- predict(xgb, Test.X)
xgb_pred
xgb_pred_class <- as.numeric(xgb_pred > 0.5) #cut off threshold는 0.5인데 이것은 0,1이 각각 나타날 확률이 0.5이기 때문임.(실제 표본의 불균형으로 0,1이 나타날 확률이 달라질 경우에는 다음 블로그와 같이 threshold 최적화 기법을 사용함:https://appliedmachinelearning.blog/2018/09/23/data-analytics-and-modeling-with-xgboost-classifier-wns-hackathon-challenge/)
confusionMatrix(as.factor(xgb_pred_class), as.factor(Test.Y),positive="1")

#변수중요도 그래프_______________________________________________________________________________________

#Determine Which Features Are Most Important

#Get the trained model
model <- xgb.dump(xgb, with_stats=TRUE)

#Get the feature real names
names <- dimnames(Train.X)[[2]]

#Compute feature importance matrix
importance_matrix <- xgb.importance(names, model=xgb)[0:20] # View top 20 most important features

#Plot
xgb.plot.importance(importance_matrix)

#ROC커브 그래프________________________________________________________________
library(ROCR)

#Use ROCR package to plot ROC Curve
xgb.pred <- prediction(xgb_pred, Test.Y)
xgb.perf <- performance(xgb.pred, "tpr", "fpr")

plot(xgb.perf,
     avg="threshold",
     colorize=TRUE,
     lwd=1,
     main="ROC Curve w/ Thresholds",
     print.cutoffs.at=seq(0, 1, by=0.05),
     text.adj=c(-0.5, 0.5),
     text.cex=0.5)
grid(col="lightgray")
axis(1, at=seq(0, 1, by=0.1))
axis(2, at=seq(0, 1, by=0.1))
abline(v=c(0.1, 0.3, 0.5, 0.7, 0.9), col="lightgray", lty="dotted")
abline(h=c(0.1, 0.3, 0.5, 0.7, 0.9), col="lightgray", lty="dotted")
lines(x=c(0, 1), y=c(0, 1), col="black", lty="dotted")
